src/ParameterDistributions.jl

```diff
@@ -0,0 +1,308 @@
+
+The number of dimensions of the parameter space
+"""
+function len(d::Parameterized) 
+    return length(d.distribution)
+end
+
+function len(d::Samples)
+    return size(d.distribution_samples)[2]
+end
+
+"""
+    struct ParameterDistribution
+
+Structure to hold a parameter distribution
+"""
+struct ParameterDistribution{PDType <: ParameterDistributionType, CType <: ConstraintType}
+    distribution::PDType
+    constraints::Array{CType}
+    name::String
+
+    ParameterDistribution(parameter_distribution::PDType,
+                          constraints::Array{CType},
+                          name::String) where {PDType <: ParameterDistributionType,
+                                               CType <: ConstraintType} = !(len(parameter_distribution) ==
+len(constraints)) ? throw(DimensionMismatch("There must be one constraint per parameter")) : new{PDType,CType}(parameter_distribution, constraints, name)     
+end
+    
+
+"""
+    struct ParameterDistributions
+
+Structure to hold an array of ParameterDistribution's
+"""
+struct ParameterDistributions
+    parameter_distributions::Array{ParameterDistribution}
+end
+
+
+
+
+## Functions
+
+"""
+    get_name(pds::ParameterDistributions)
+
+Returns a list of ParameterDistribution names
+"""
+function get_name(pds::ParameterDistributions)
+    return [get_name(pd) for pd in pds.parameter_distributions]
+end
+
+function get_name(pd::ParameterDistribution)
+    return pd.name
+end
+
+"""
+    get_distributions(pds::ParameterDistributions)
+
+Returns a `Dict` of `ParameterDistribution` distributions by name, (unless sample type)
+"""
+function get_distribution(pds::ParameterDistributions)
+    return Dict{String,Any}(get_name(pd) => get_distribution(pd) for pd in pds.parameter_distributions)
+end
+
+function get_distribution(pd::ParameterDistribution)
+    return get_distribution(pd.distribution)
+end
+
+function get_distribution(d::Samples)
+    return "Contains samples only"
+end
+function get_distribution(d::Parameterized)
+    return d.distribution
+end
+
+"""
+    function sample_distribution(pds::ParameterDistributions)
+
+Draws samples from the parameter distributions
+"""
+function sample_distribution(pds::ParameterDistributions)
+    return sample_distribution(pds,1)
+end
+
+function sample_distribution(pds::ParameterDistributions, n_samples::IT) where {IT <: Integer}
+    return Dict{String,Any}(get_name(pd) => sample_distribution(pd, n_samples) for pd in pds.parameter_distributions)
+end
+
+function sample_distribution(pd::ParameterDistribution)
+    return sample_distribution(pd,1)
+end
+
+function sample_distribution(pd::ParameterDistribution, n_samples::IT) where {IT <: Integer}
+    return sample_distribution(pd.distribution,n_samples)
+end
+
+function sample_distribution(d::Samples, n_samples::IT)  where {IT <: Integer}
+    total_samples = size(d.distribution_samples)[1]
+    samples_idx = StatsBase.sample(collect(1:total_samples), n_samples; replace=false)
```

Comment:
Should sampling be without replacement?

Reply:
I would say sampling should be with replacement. Then, this becomes the empirical distribution function of the sample space. This also lends itself to the use of bootstrap for estimating sample statistics.